torch/_logging/_internal.py

```diff
@@ -0,0 +1,359 @@
+import logging
+import functools
+import os
+import re
+from importlib import __import__
+from typing import Dict, Set
+from dataclasses import dataclass, field
+
+DEFAULT_LOG_LEVEL = logging.WARN
+DEFAULT_FORMATTER = logging.Formatter(
+    "[%(asctime)s] %(name)s: [%(levelname)s] %(message)s"
+)
+LOG_ENV_VAR = "TORCH_LOGS"
+
+
+@dataclass
+class LogRegistry:
+    # shorthand name to log qualified name
+    log_alias_to_log_qname : Dict[str, str] = field(default_factory=dict)
+
+    artifact_log_qnames : Set[str] = field(default_factory=set)
+
+    artifact_names : Set[str] = field(default_factory=set)
+
+    # shorthand names of logs which support verbosity
+    # e.g. AOT only has debug messages, so INFO level is useless
+    log_aliases_with_verbosity: Set[str] = field(default_factory=set)
+
+    off_by_default_artifact_names: Set[str] = field(default_factory=set)
+
+    def is_artifact(self, name):
+        return name in self.artifact_names
+
+    def is_log(self, alias):
+        return alias in self.log_alias_to_log_qname
+
+    def register_log(self, alias, log_qname, has_verbosity):
+        self.log_alias_to_log_qname[alias] = log_qname
+
+        if has_verbosity:
+            self.log_aliases_with_verbosity.add(alias)
+
+    def register_artifact_name(self, name, off_by_default):
+        self.artifact_names.add(name)
+
+        # if off by default, don't enable it
+        # when log_name's log_level is set to DEBUG
+        if off_by_default:
+            self.off_by_default_artifact_names.add(name)
+
+    def register_artifact_log(self, artifact_log_qname):
+        self.artifact_log_qnames.add(artifact_log_qname)
+
+    def get_log_qnames(self):
+        return set(self.log_alias_to_log_qname.values())
+
+    def get_artifact_log_qnames(self):
+        return set(self.artifact_log_qnames)
+
+    def supports_verbosity(self, log_alias):
+        return log_alias in self.log_aliases_with_verbosity
+
+    def is_off_by_default(self, artifact_qname):
+        return artifact_qname in self.off_by_default_artifact_names
+
+@dataclass
+class LogState:
+    log_qname_to_level : Dict[str, str] = field(default_factory=dict)
+
+    artifact_names : set[str] = field(default_factory=set)
+
+    def enable_artifact(self, artifact_name):
+        self.artifact_names.add(artifact_name)
+
+    def is_artifact_enabled(self, name):
+        return name in self.artifact_names
+
+    def enable_log(self, log_qname, log_level):
+        self.log_qname_to_level[log_qname] = log_level
+
+    def get_log_level_pairs(self):
+        return list(self.log_qname_to_level.items())
```

Comment:
nit: there isn't really any reason to reify this into a list; the only use site iterates over it once